artifact.go

```diff
@@ -163,50 +151,311 @@ func (m ArtifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+	buildID, files, ok := parseBuildIDAndFiles(c)
+	if !ok {
+		return
+	}
+
+	for _, f := range files {
+		artifact := Artifact{
+			Data:     f.data,
+			Name:     f.name,
+			FileName: f.fileName,
+			BuildID:  buildID,
+		}
+
+		err := m.Database.
+			Create(&artifact).
+			Error
+		if err != nil {
+			ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+				"Failed saving artifact with name %q for build with ID %d in database.",
+				artifact.FileName, buildID))
+			return
+		}
+
+		log.Info().
+			WithString("filename", artifact.Name).
+			WithUint("build", buildID).
+			WithUint("artifact", artifact.ArtifactID).
+			Message("File saved as artifact")
+	}
+
+	c.Status(http.StatusCreated)
+}
+
+type file struct {
+	name     string
+	fileName string
+	data     []byte
+}
+
+// putTestResultDataHandler godoc
+// @summary Post test result data
+// @tags artifact
+// @accept multipart/form-data
+// @param buildid path int true "Build ID"
+// @param file formData file true "Test result artifact file"
+// @success 201 {object} SummaryOfTestResultSummaries "Added new test result data and created summary"
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result-data [put]
+func (m ArtifactModule) putTestResultDataHandler(c *gin.Context) {
+	buildID, files, ok := parseBuildIDAndFiles(c)
+	if !ok {
+		return
+	}
+
+	artifacts := []*Artifact{}
+	for _, f := range files {
+		artifact := &Artifact{
+			Data:     f.data,
+			Name:     f.name,
+			FileName: f.fileName,
+			BuildID:  buildID,
+		}
+
+		err := m.Database.Create(artifact).Error
+		if err != nil {
+			ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+				"Failed saving artifact with name %q for build with ID %d in database.",
+				artifact.FileName, buildID))
+			return
+		}
+
+		artifacts = append(artifacts, artifact)
+
+		log.Debug().
+			WithString("filename", artifact.Name).
+			WithUint("build", buildID).
+			WithUint("artifact", artifact.ArtifactID).
+			Message("File saved as artifact")
+	}
+
+	summaries := []*TestResultSummary{}
+	details := []*TestResultDetail{}
+
+	for _, artifact := range artifacts {
+		detail, summary, err := parseAsTRX(artifact.Data)
+		if err != nil {
+			log.Warn().
+				WithError(err).
+				WithString("filename", artifact.Name).
+				WithUint("build", buildID).
+				WithUint("artifact", artifact.ArtifactID).
+				Message("Failed to unmarshal; invalid JSON format")
+			continue
+		}
+
+		for _, d := range detail {
+			d.ArtifactID = artifact.ArtifactID
+			d.BuildID = buildID
+		}
+		summary.ArtifactID = artifact.ArtifactID
+		summary.FileName = artifact.FileName
+		summary.Artifact = artifact
+		summary.BuildID = buildID
+
+		summaries = append(summaries, summary)
+		details = append(details, detail...)
+	}
+
+	summaryOfSummaries := SummaryOfTestResultSummaries{
+		Failed:    0,
+		Passed:    0,
+		Skipped:   0,
+		Total:     0,
```

Comment:
These defaults to zero. Think it will be equally readable if you just omitted these declarations